x-pack/plugins/security/server/authentication/providers/anonymous.ts

```diff
@@ -0,0 +1,217 @@
+import { AuthenticationResult } from '../authentication_result';
+import { canRedirectRequest } from '../can_redirect_request';
+import { DeauthenticationResult } from '../deauthentication_result';
+import {
+  BasicHTTPAuthorizationHeaderCredentials,
+  HTTPAuthorizationHeader,
+} from '../http_authentication';
+import { AuthenticationProviderOptions, BaseAuthenticationProvider } from './base';
+
+/**
+ * Credentials that are based on the username and password.
+ */
+interface UsernameAndPasswordCredentials {
+  username: string;
+  password: string;
+}
+
+/**
+ * Credentials that are based on the Elasticsearch API key with the optional username override.
+ */
+interface APIKeyCredentials {
+  username?: string;
+  apiKey: string;
+}
+
+/**
+ * Checks whether current request can initiate a new session.
+ * @param request Request instance.
+ */
+function canStartNewSession(request: KibanaRequest) {
+  // We should try to establish new session only if request requires authentication and it's not XHR request.
+  // Technically we can authenticate XHR requests too, but we don't want these to create a new session unintentionally.
+  return canRedirectRequest(request) && request.route.options.authRequired === true;
+}
+
+/**
+ * Checks whether specified `credentials` define an API key.
+ * @param credentials
+ */
+function isAPIKeyCredentials(
+  credentials: UsernameAndPasswordCredentials | APIKeyCredentials
+): credentials is APIKeyCredentials {
+  const apiKey = ((credentials as unknown) as Record<string, any>).apiKey;
+  return typeof apiKey === 'string' && !!apiKey;
+}
+
+/**
+ * Provider that supports anonymous request authentication.
+ */
+export class AnonymousAuthenticationProvider extends BaseAuthenticationProvider {
+  /**
+   * Type of the provider.
+   */
+  static readonly type = 'anonymous';
+
+  /**
+   * Specifies credentials that should be used to authenticate anonymous user.
+   */
+  private readonly credentials: Readonly<UsernameAndPasswordCredentials | APIKeyCredentials>;
+
+  constructor(
+    protected readonly options: Readonly<AuthenticationProviderOptions>,
+    anonymousOptions?: Readonly<{
+      credentials?: Readonly<UsernameAndPasswordCredentials | APIKeyCredentials>;
+    }>
+  ) {
+    super(options);
+
+    if (!anonymousOptions || !anonymousOptions.credentials) {
+      throw new Error('Credentials must be specified');
+    }
+    this.credentials = anonymousOptions.credentials;
+  }
+
+  /**
+   * Performs initial login request.
+   * @param request Request instance.
+   * @param state Optional state value previously stored by the provider.
+   */
+  public async login(request: KibanaRequest, state?: unknown) {
+    this.logger.debug('Trying to perform a login.');
+
+    // If authentication succeeded we should return some state to create a session, that will reuse for all subsequent
+    // requests and anonymous user interactions with the Kibana.
+    return isAPIKeyCredentials(this.credentials)
```

Comment:
First of all, I definitely agree that we should build a safety net around anonymous access!

But I was thinking about this as a part of a broader initiative around the security center where we could display things like this to the users who are actually authorized to act on them (e.g. administrators). If we offload that work to the security center or alike we can implement more sophisticated heuristics since even though `manage_security` is dangerous it's not the only thing that can do harm.

I'm also hesitant to do that additional check for every request that requires authentication for performance reasons knowing that it may not actually prevent damage in the end...

I tend to move discussion of the possible ideas and solutions to the followup. Does that sound good to you?

Reply:
> I tend to move discussion of the possible ideas and solutions to the followup. Does that sound good to you?

tl;dr Yes, let's take this to a followup!


> But I was thinking about this as a part of a broader initiative around the security center where we could display things like this to the users who are actually authorized to act on them (e.g. administrators). If we offload that work to the security center or alike we can implement more sophisticated heuristics since even though manage_security is dangerous it's not the only thing that can do harm.

I certainly see the benefit of including these types of checks within the security center. Keep in mind though that a powerful anonymous user could choose to dismiss these checks so the real administrator might not notice.

> I'm also hesitant to do that additional check for every request that requires authentication for performance reasons knowing that it may not actually prevent damage in the end...

I agree - I was thinking we could only check on `login`, rather than on every call to `authenticate`. It's less than perfect, but would give us reasonable protection.

Reply:
> Keep in mind though that a powerful anonymous user could choose to dismiss these checks so the real administrator might not notice.

That's a good point. We can also design security center in a way that important actions/warnings won't go unnoticed by others in the admin group though, but it's an orthogonal topic anyway.

> I agree - I was thinking we could only check on login, rather than on every call to authenticate. It's less than perfect, but would give us reasonable protection.

I see, let's discuss pros and cons in the follow-up then.